src/Npgsql/PostgresException.cs

```diff
@@ -31,14 +29,206 @@ public sealed class PostgresException : NpgsqlException
+                case PostgresErrorCodes.IoError:
+                case PostgresErrorCodes.SerializationFailure:
+                case PostgresErrorCodes.LockNotAvailable:
+                case PostgresErrorCodes.ObjectInUse:
+                case PostgresErrorCodes.ObjectNotInPrerequisiteState:
+                case PostgresErrorCodes.ConnectionException:
+                case PostgresErrorCodes.ConnectionDoesNotExist:
+                case PostgresErrorCodes.ConnectionFailure:
+                case PostgresErrorCodes.SqlClientUnableToEstablishSqlConnection:
+                case PostgresErrorCodes.SqlServerRejectedEstablishmentOfSqlConnection:
+                case PostgresErrorCodes.TransactionResolutionUnknown:
+                    return true;
+                default:
+                    return false;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the statement which triggered this exception.
+        /// </summary>
+        public NpgsqlStatement Statement { get; internal set; }
+
+        /// <summary>
+        /// Gets a collection of key/value pairs that provide additional PostgreSQL fields about the exception.
+        /// </summary>
+        public override IDictionary Data
+        {
+            get
+            {
+                var data = base.Data;
+                if (_dataInitialized)
+                    return data;
+                
+                AddData(nameof(Severity), Severity);
+                AddData(nameof(SqlState), SqlState);
+                AddData(nameof(MessageText), MessageText);
+                AddData(nameof(Detail), Detail);
+                AddData(nameof(Hint), Hint);
+                AddData(nameof(Position), Position);
+                AddData(nameof(InternalPosition), InternalPosition);
+                AddData(nameof(InternalQuery), InternalQuery);
+                AddData(nameof(Where), Where);
+                AddData(nameof(SchemaName), SchemaName);
+                AddData(nameof(TableName), TableName);
+                AddData(nameof(ColumnName), ColumnName);
+                AddData(nameof(DataTypeName), DataTypeName);
+                AddData(nameof(ConstraintName), ConstraintName);
+                AddData(nameof(File), File);
+                AddData(nameof(Line), Line);
+                AddData(nameof(Routine), Routine);
+
+                _dataInitialized = true;
+                return data;
+
+                void AddData<T>(string key, T value)
+                {
+                    if (!EqualityComparer<T>.Default.Equals(value, default))
+                        data.Add(key, value);
+                }
+            }
+        }
+
         #region Message Fields
 
         /// <summary>
         /// Severity of the error or notice.
         /// Always present.
         /// </summary>
         [PublicAPI]
-        public string Severity { get; set; }
+        public string Severity { get; }
```

Comment:
I don't remember anymore whether there was a specific reason we had a public setter here... In any case I'm OK with removing this.